votor/src/certificate_pool.rs

```diff
@@ -87,6 +87,105 @@ impl From<AlpenglowCommitmentError> for AddVoteError {
     }
 }
 
+#[derive(Debug, Default)]
+struct FallbackVoteCounters {
+    my_first_vote: Option<Vote>,
+    total_stake: Stake,
+    skip_total: Stake,
+    notarize_total: Stake,
+    notarize_entry_total: BTreeMap<Hash, Stake>,
+    top_notarized_stake: Stake,
+    safe_to_notar_sent: Vec<Hash>,
+    safe_to_skip_sent: bool,
+}
+
+impl FallbackVoteCounters {
+    pub fn add_vote(
+        &mut self,
+        vote: &Vote,
+        total_stake: Stake,
+        entry_stake: Stake,
+        is_my_own_vote: bool,
+        events: &mut Vec<VotorEvent>,
+    ) -> (usize, bool) {
+        match vote {
+            Vote::Skip(_) => self.skip_total = entry_stake,
+            Vote::Notarize(vote) => {
+                self.notarize_entry_total
+                    .insert(*vote.block_id(), entry_stake);
+                self.notarize_total = self.notarize_entry_total.values().sum();
+                self.top_notarized_stake = self.top_notarized_stake.max(entry_stake);
+            }
+            _ => return (0, false), // Not interested in other vote types
+        }
+        if self.my_first_vote.is_none() && is_my_own_vote {
+            self.my_first_vote = Some(*vote);
+            self.total_stake = total_stake;
+        }
+        if self.my_first_vote.is_some() {
```

Comment:
we can make this even more granular, e.g. only check `is_safe_to_skip` if `my_first_vote` was a notarize. Then we could also remove the check in the function.

I don't think we need to keep track of `my_first_vote: Vote` fully, could have an enum with `my_vote : Notarize | Skip | None`

Reply:
Err, we do need to check the block_id in Notarize in safe_to_notar. I don't think saving Vote instead of VoteType saves us that much any way.

I can move check on my_first_vote out, but safe_to_notar also need to check my_first_vote, so it's nice to keep the check here, and we can keep comments in the paper in one location. I'm leaning a bit towards checking inside the function, because I can put the comments in the function, but I can be convinced.